src/storage/src/sst/pruning.rs

```diff
@@ -0,0 +1,320 @@
+use parquet::arrow::arrow_reader::{ArrowPredicate, RowFilter};
+use parquet::arrow::ProjectionMask;
+use parquet::schema::types::SchemaDescriptor;
+use table::predicate::Predicate;
+
+use crate::error;
+use crate::schema::StoreSchema;
+
+/// Builds time range row filter.
+pub(crate) fn build_row_filter(
+    time_range: TimestampRange,
+    predicate: &Predicate,
+    store_schema: &Arc<StoreSchema>,
+    schema_desc: &SchemaDescriptor,
+    projection_mask: ProjectionMask,
+) -> Option<RowFilter> {
+    let ts_col_idx = store_schema.timestamp_index();
+    let ts_col = store_schema.columns().get(ts_col_idx)?;
+    let ts_col_unit = match &ts_col.desc.data_type {
+        ConcreteDataType::Int64(_) => TimeUnit::Millisecond,
+        ConcreteDataType::Timestamp(ts_type) => ts_type.unit(),
+        _ => unreachable!(),
+    };
+
+    let ts_col_projection = ProjectionMask::roots(schema_desc, vec![ts_col_idx]);
+
+    // checks if converting time range unit into ts col unit will result into rounding error.
+    if time_unit_lossy(&time_range, ts_col_unit) {
+        let filter = RowFilter::new(vec![Box::new(PlainTimestampRowFilter::new(
+            time_range,
+            ts_col_projection,
+        ))]);
+        return Some(filter);
+    }
+
+    // If any of the conversion overflows, we cannot use arrow's computation method, instead
+    // we resort to plain filter that compares timestamp with given range, less efficient,
+    // but simpler.
+    // TODO(hl): If the range is gt_eq/lt, we also use PlainTimestampRowFilter, but these cases
+    // can also use arrow's gt_eq_scalar/lt_scalar methods.
+    let time_range_row_filter = if let (Some(lower), Some(upper)) = (
+        time_range
+            .start()
+            .and_then(|s| s.convert_to(ts_col_unit))
+            .map(|t| t.value()),
+        time_range
+            .end()
+            .and_then(|s| s.convert_to(ts_col_unit))
+            .map(|t| t.value()),
+    ) {
+        Box::new(FastTimestampRowFilter::new(ts_col_projection, lower, upper)) as _
+    } else {
+        Box::new(PlainTimestampRowFilter::new(time_range, ts_col_projection)) as _
+    };
+    let mut predicates = vec![time_range_row_filter];
+    if let Ok(datafusion_filters) = predicate_to_row_filter(predicate, projection_mask) {
+        predicates.extend(datafusion_filters);
+    }
+    let filter = RowFilter::new(predicates);
+    Some(filter)
+}
+
+fn predicate_to_row_filter(
+    predicate: &Predicate,
+    projection_mask: ProjectionMask,
+) -> error::Result<Vec<Box<dyn ArrowPredicate>>> {
+    let mut datafusion_predicates = Vec::with_capacity(predicate.exprs().len());
+    for expr in predicate.exprs() {
+        datafusion_predicates.push(Box::new(DatafusionArrowPredicate {
+            projection_mask: projection_mask.clone(),
+            physical_expr: expr.clone(),
+        }) as _);
+    }
+
+    Ok(datafusion_predicates)
+}
+
+#[derive(Debug)]
+struct DatafusionArrowPredicate {
```

Comment:
Does datafusion provide similar struct so we can reuse it?

Reply:
datafusion does have a similar struct but it's private to `datafusion::physical_plan`.